solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -0,0 +1,155 @@
+"""
+
+import pandas as pd
+
+from solarforecastarbiter import datamodel
+from solarforecastarbiter.validation import quality_mapping
+
+
+def apply_validation(data, qfilter, handle_func):
+    """
+    Apply validation steps based on provided filters to the data.
+
+    Parameters
+    ----------
+    qfilter : list of
+        :class:`solarforecastarbiter.datamodel.QualityFlagFilter`
+    data : :class:`pd.DataFrame`
+        Pandas DataFrame of observations and forecasts.
+    handle_func : function
+        Function that handles how `quality_flags` will be used.
+
+    Returns
+    -------
+    pd.DataFrame
+    """
+    validated_data = {}
+
+    # List of flags from filter
+    if not isinstance(qfilter, datamodel.QualityFlagFilter):
+        return TypeError(f"{qfilter} not a QualityFlagFilter")
+    filters = qfilter.quality_flags
+
+    # Apply handling function to quality flags
+    for model, values in data.items():
+
+        # Skip if empty
+        if isinstance(values, pd.DataFrame) and values.empty:
+            validated_data[model] = values.value
+            continue
+
+        # Apply only to Observations
+        if isinstance(model, datamodel.Observation):
+            validation_df = quality_mapping.convert_mask_into_dataframe(
+                values['quality_flag'])
+            validation_df = validation_df[filters]
+            validated_data[model] = handle_func(values.value,
+                                                validation_df.any(axis=1))
+        elif isinstance(model, datamodel.Forecast):
+            validated_data[model] = values
+        else:
+            raise TypeError(f"{model} not an Observation or Forecast")
+
+    return validated_data
+
+
+def resample_and_align(fx_obs, data, tz):
+    """
+    Resample to the forecast and observation using the the larger interval
+    length  and align them to overlap.
+
+    Parameters
+    ----------
+    fx_obs : :class:`solarforecastarbiter.datamodel.ForecastObservation`
+        Pair of forecast and observation.
+    data : dict
+        Dictionary that must include the timeseries of the pair.
+    tz : str
+        Timezone to witch processed data will be converted.
+
+    Returns
+    -------
+    :class:`solarforecastarbiter.datamodel.ProcessedForecastObservation`
+
+    Todo
+    ----
+      * Add ability to set use smaller interval length
+      * Add other resampling functions (besides mean like first, last, median)
+    """
+    fx = fx_obs.forecast
+    obs = fx_obs.observation
+
+    # Check that the forecast and observation models exist in the data
+    assert fx in data.keys(), f"Forecast not found in data."
+    assert obs in data.keys(), f"Observation not found in data."
+
+    # Get minimum and maximum datetimes???
+    # min_dt = max(fx_data.index.min(), obs_data.index.min())
+    # max_dt = min(fx_data.index.max(), obs_data.index.max())
+
+    # Determine larger interval length
+    if fx.interval_length >= obs.interval_length:
```

Comment:
Good point. I didn't provide this test case and will add it.

This brings up another good point. Could it ever be the case that the reference forecast needs to be resampled?  I'm thinking maybe I need to add an option to the function (not exposed to the reporting) to provide the resample interval as an argument.  Do you agree?

Reply:
Maybe. Right now I think we'd want to create a new reference forecast for the application, but I'm not certain that this would always be the case. What do you think @alorenzo175?

Reply:
Just realized that the datamodel.ForecastObservation object does not allow it:

here is the error
```
    def __check_interval_compatibility__(forecast, observation):
        if observation.interval_length > forecast.interval_length:
>           raise ValueError('observation.interval_length cannot be greater than '
                             'forecast.interval_length.')
E           ValueError: observation.interval_length cannot be greater than forecast.interval_length.

solarforecastarbiter/datamodel.py:645: ValueError
```

I think I should just go ahead and remove support for observation interval length being less than forecast since it can't occur according to the model.

Reply:
I think it's good to never resample forecasts, including reference forecasts. So, yes we would create a new reference forecast if we needed a shorter interval for some application.
> I think I should just go ahead and remove support for observation interval length being less than forecast since it can't occur according to the model.

Probably just a typo, but you should remove support for *forecast* interval length being less than *observation* interval length

Reply:
I can add that if you want. Just let me know.

Reply:
@alorenzo175 Yes that was a mixup. I've removed support for it now.

In the #224 I will create a new/temporary ForecastObservation object that won't persist outside the function and has resampled reference forecast that matches the original forecast if the interval lengths don't match (and of course it is provided at all).